File: core/src/main/java/ru/vertuos/graphics/hedgehog/listener/SonicStateListener.java
package ru.vertuos.graphics.hedgehog.listener;

import com.badlogic.gdx.graphics.g2d.TextureRegion;
import ru.vertuos.engine.hedgehog.Hedgehog;
import ru.vertuos.engine.hedgehog.listener.HedgehogPropertyChangeListener;
import ru.vertuos.engine.world.object.DynamicGameObject;
import ru.vertuos.ui.animation.contracts.HedgehogAnimationContract;
import ru.vertuos.ui.animation.manager.AnimationManager;

public class SonicStateListener implements HedgehogPropertyChangeListener {

    private final AnimationManager<TextureRegion> manager;

    public SonicStateListener(AnimationManager<TextureRegion> manager) {
        this.manager = manager;
    }

    @Override
    public void onStateChange(Hedgehog.State state) {
        switch (state) {
            case IDLE:
                manager.selectAnimation(HedgehogAnimationContract.KEY_IDLE);
                break;
            case WALK:
                manager.selectAnimation(HedgehogAnimationContract.KEY_WALK);
                break;
            case RUN:
                manager.selectAnimation(HedgehogAnimationContract.KEY_RUN);
                break;
            case SPIN:
                manager.selectAnimation(HedgehogAnimationContract.KEY_SPIN);
                break;
            case SPINDASH:
                manager.selectAnimation(HedgehogAnimationContract.KEY_SPINDASH);
        }
    }

    @Override
    public void onPositionChange(float newPosX, float newPosY) {

    }

    @Override
    public void onLinearVelocityChange(float newVelX, float newVelY) {

    }

    @Override
    public void onAccelerationChange(float newAccX, float newAccY) {

    }

    @Override
    public void onHpChange(float newHp) {

    }

    @Override
    public void onLivesCountChange(int newLivesCount) {

    }

    @Override
    public void onRingsCountChange(int newRingsCount) {

    }

    @Override
    public void onOnGroundChange(boolean isOnGroundNow) {

    }

    @Override
    public void onChangeDirection(DynamicGameObject.Direction dir) {
        manager.setFlipped(dir == DynamicGameObject.Direction.BACKWARD);
    }
}

File: core/src/main/java/ru/vertuos/graphics/hedgehog/listener/SonicStateListener.kt
package ru.vertuos.graphics.hedgehog.listener

import com.badlogic.gdx.graphics.g2d.TextureRegion
import ru.vertuos.engine.hedgehog.Hedgehog
import ru.vertuos.engine.hedgehog.listener.HedgehogPropertyChangeListener
import ru.vertuos.engine.world.obj.DynamicGameObject.Direction
import ru.vertuos.ui.animation.contracts.HedgehogAnimationContract
import ru.vertuos.ui.animation.manager.AnimationManager

class SonicStateListener(
    private val manager: AnimationManager<TextureRegion>
) : HedgehogPropertyChangeListener {
    override fun onStateChange(newState: Hedgehog.State) {
        when (newState) {
            Hedgehog.State.IDLE -> manager.selectAnimation(HedgehogAnimationContract.KEY_IDLE)
            Hedgehog.State.WALK -> manager.selectAnimation(HedgehogAnimationContract.KEY_WALK)
            Hedgehog.State.RUN -> manager.selectAnimation(HedgehogAnimationContract.KEY_RUN)
            Hedgehog.State.SPIN -> manager.selectAnimation(HedgehogAnimationContract.KEY_SPIN)
            Hedgehog.State.SPINDASH -> manager.selectAnimation(HedgehogAnimationContract.KEY_SPINDASH)
            Hedgehog.State.DUCK -> Unit
            Hedgehog.State.LOOK_UP -> Unit
        }
    }

    override fun onPositionChange(newPosX: Float, newPosY: Float) {
    }

    override fun onLinearVelocityChange(newVelX: Float, newVelY: Float) {
    }

    override fun onAccelerationChange(newAccX: Float, newAccY: Float) {
    }

    override fun onHpChange(newHp: Float) {
    }

    override fun onLivesCountChange(newLivesCount: Int) {
    }

    override fun onRingsCountChange(newRingsCount: Int) {
    }

    override fun onOnGroundChange(isOnGroundNow: Boolean) {
    }

    override fun onChangeDirection(dir: Direction) {
        manager.isFlipped = dir == Direction.BACKWARD
    }
}
